client/src/components/mapping/ConceptMap/index.tsx

```diff
@@ -0,0 +1,569 @@
+  };
+
+  const fetchCodesForCodeSystem = (systemName: string): string[] =>
+    existingCodeSystems
+      .find(s => s.title === systemName)
+      ?.concept.map(c => c.code) || [];
+
+  const selectSourceCodeSystem = (
+    <CodeSystemSelect
+      systems={existingCodeSystems}
+      selectedSystem={sourceSystemName}
+      itemDisabled={isSourceSystemDisabled}
+      onChange={(s: string): void => {
+        if (s !== sourceSystemName) {
+          setSourceSystemName(s);
+          resetMap();
+        }
+      }}
+      onClear={(): void => {
+        setSourceSystemName('');
+        setCreatingNewCodeSystem(false);
+        resetMap();
+      }}
+      allowCreate={true}
+      callbackCreatingNewSystem={(): void => {
+        resetMap();
+        setSourceSystemName('');
+        setCreatingNewCodeSystem(true);
+      }}
+    />
+  );
+
+  const enterNewCodeSystemName = (
+    <div className="enter-code-name">
+      <input
+        className="text-input"
+        placeholder="Enter name..."
+        value={sourceSystemName}
+        type="text"
+        onChange={(e: React.ChangeEvent<HTMLInputElement>): void =>
+          setSourceSystemName(e.target.value)
+        }
+      />
+      <Button
+        icon="cross"
+        minimal={true}
+        onClick={(): void => {
+          setCreatingNewCodeSystem(false);
+          setSourceSystemName('');
+        }}
+      />
+    </div>
+  );
+
+  const selectTargetCodeSystem = (
+    <CodeSystemSelect
+      systems={existingCodeSystems}
+      selectedSystem={selectedTargetSystem}
+      itemDisabled={isTargetSystemDisabled}
+      onChange={(s: string): void => {
+        if (s !== selectedTargetSystem) {
+          setSelectedTargetSystem(s);
+          resetMap();
+        }
+      }}
+      onClear={(): void => {
+        setSelectedTargetSystem('');
+        resetMap();
+      }}
+    />
+  );
+
+  const createSourceCode = (index: number) => (
+    <input
+      className="text-input"
+      key={index}
+      value={conceptMap[index].source}
+      type="text"
+      onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
+        setConceptMap(prev => {
+          prev[index].source = e.target.value;
+          return [...prev];
+        })
+      }
+    />
+  );
+
+  const deleteRowButton = (index: number) => (
+    <Button
+      icon="trash"
+      minimal={true}
+      onClick={(): void => {
+        setConceptMap(prev => {
+          prev.splice(index, 1);
+          return [...prev];
+        });
+      }}
+    />
+  );
+
+  const chooseSourceCode = (index: number) => (
+    <StringSelect
+      inputItem={conceptMap[index].source}
+      items={fetchCodesForCodeSystem(sourceSystemName)}
+      displayItem={displayCode}
+      onChange={(code: string): void =>
+        setConceptMap(prev => {
+          prev[index].source = code;
+          return [...prev];
+        })
+      }
+    />
+  );
+
+  const chooseTargetCode = (index: number) => (
```

Comment:
can you factorize with the prev function ? you can pass an argument of type `"source" | "target"` to the function for instance

Reply:
Done